### PR TITLE
Remove limitation on the maximum number of concurrent connections handled by spiped

### PR DESCRIPTION
**Pull request by cperciva**

### CODE REVIEW

File: spiped/main.c [hunk 1]
```diff
@@ -253,6 +253,14 @@ main(int argc, char * argv[])
 	if (opt_t == NULL)
 		usage();
 
+	/*
+	 * A limit of SIZE_MAX connections is equivalent to any larger limit;
+	 * we'll be unable to allocate memory for socket bookkeeping before we
+	 * reach either.
+	 */
+	if (opt_n > SIZE_MAX)
```
Comment (gperciva) [inline spiped/main.c]: My patch did this testing in the `GETOPT_OPTARG` block.

**Comment by gperciva** [comment]
For the record, the log for the failing build looks normal to me.  It should go away once we have #110 and the other 2 related PRs merged.

**Comment by cperciva** [comment]
Which are "the other 2 related PRs"?

**Comment by gperciva** [comment]
I resubmitted this PR to get it testing, but I noticed a few compiler warnings.

File: spiped/main.c [hunk 2]
```diff
@@ -134,16 +135,15 @@ main(int argc, char * argv[])
 			opt_k = optarg;
 			break;
 		GETOPT_OPTARG("-n"):
-			if (opt_n != 0)
+			if (opt_n_set)
 				usage();
+			opt_n_set = 1;
 			if ((opt_n = strtoimax(optarg, NULL, 0)) == 0) {
```
Comment (gperciva) [inline spiped/main.c]: This does not permit the user to set `-n 0`.

Unfortunately, we can't rely on errno; "If no conversion could be performed, zero shall be returned [CX] ￼  and errno may be set to [EINVAL]. "
http://pubs.opengroup.org/onlinepubs/9699919799/functions/strtoimax.html

The man page on my Ubuntu machine says "If nothing was  found  to convert, zero is returned.  On overflow or underflow INTMAX_MAX or INTMAX_MIN or UINTMAX_MAX is returned, and errno is set to ERANGE." so POSIX wasn't kidding when they wrote that *errno* **may** be set.

File: spiped/spiped.1 [hunk 3]
```diff
@@ -120,7 +120,11 @@ Disable transport layer keep-alives.
 .TP
 .B \-n <max # connections>
 Limit on the number of simultaneous connections allowed.
-This value must be between 1 and 500.
+A value of 0 indicates that no limit should be imposed; this may be
```
Comment (gperciva) [inline spiped/spiped.1]: Given the (astonishing) difficulty of parsing `-n 0`, and the fact that we don't actually make this unlimited (as it's still bound by `SIZE_MAX`), why don't we remove this "value of 0" stuff?  If somebody wants a (virtually) unlimited number of connections, they can do `-n 1234567890`.  I doubt that anybody will have more than 2^30 connections.

**Comment by cperciva** [comment]
Note to self: Blocked on https://github.com/Tarsnap/libcperciva/pull/54 or some variant thereof, which we'll use to tidy up the handling of the `-n` option (and also others).

**Comment by cperciva** [comment]
@gperciva can you merge PARSENUM to spiped do whatever git magic is needed so that this branch is forked from after that import?  I'd like to get this into the tree soon.

**Comment by gperciva** [comment]
Yes, once we've fixed the `INFINITY` thing.

**Comment by gperciva** [comment]
WARNING updated PR and rewrote your commit.

Due to merge conflicts, I had to edit the existing commit.  At first I tried to obsessively retain the code changes you made, and put all modifications in a separate commit, but then I figured that since this is a relatively easy change and I had to do *some* edits anyway, I might as well just rewrite the "Remove limit on spiped -n flag" commit.

This PR now includes #124, as that would have added another merge conflict.  In theory, we should be able to:
1) merge #124.
2) refresh this PR.  (hopefully just a browser refresh, but potentially another force-push but without changing any code)
3) merge this PR.

File: spiped/main.c [hunk 4]
```diff
@@ -71,7 +71,8 @@ main(int argc, char * argv[])
 	int opt_F = 0;
 	int opt_j = 0;
 	const char * opt_k = NULL;
-	intmax_t opt_n = 0;
+	int opt_n_set = 0;
+	uintmax_t opt_n = 0;
```
Comment (gperciva) [inline spiped/main.c]: if we still have `intmax_t opt_n`, then clang complains with:

```
main.c:149:13: error: implicit conversion changes signedness: 'unsigned long' to
      'intmax_t' (aka 'long') [-Werror,-Wsign-conversion]
                                opt_n = SIZE_MAX;
                                      ~ ^~~~~~~~
```
(amongst other similar complaints)

Comment (cperciva) [inline spiped/main.c]: Can't this change to be a size_t now?

File: spiped/main.c [hunk 5]
```diff
@@ -137,12 +138,15 @@ main(int argc, char * argv[])
...
-			if (opt_n != 0)
+			if (opt_n_set)
 				usage();
-			if (PARSENUM(&opt_n, optarg, 1, 500)) {
-				warn0("The parameter to -n must be between 1 and 500\n");
+			if (PARSENUM(&opt_n, optarg)) {
+				warn0("Invalid option: -n %s", optarg);
 				exit(1);
 			}
+			opt_n_set = 1;
```
Comment (cperciva) [inline spiped/main.c]: Move up to the line after `usage()` match style elsewhere.

**Comment by gperciva** [comment]
Updated PR, including:
- `size_t opt_n`
- `opt_n_set = 1;` after the `if`
- removed unnecessary `(size_t)opt_n` cast

**Pull request merged**